File: templating/tag_contributions/src/main/java/com/vmware/jsonteng/ext/tags/net/Ipv4HostIpTag.java
package com.vmware.jsonteng.ext.tags.net;

import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;

import com.vmware.jsonteng.ElementResolver;
import com.vmware.jsonteng.TagResolver;
import com.vmware.jsonteng.TemplateEngineException;
import com.vmware.jsonteng.tags.TagBase;

public class Ipv4HostIpTag extends TagBase {
    public static final String name = "ipv4-host-ip";

    private final ElementResolver elementResolver;

    public Ipv4HostIpTag(TagResolver tagResolver) {
        super(tagResolver);
        elementResolver = tagResolver.getElementResolver();
    }

    @Override
    public Object process(List<?> tagTokens, List<Map<String, ?>> bindingDataList) throws TemplateEngineException {
        int tokenCount = tagTokens.size();
        if (tokenCount < 2) {
            throw new TemplateEngineException(
                    String.format("Tag \"%s\" needs at least 2 parameters. Parameters given %s", name, tagTokens));
        }
        String resolvedNetwork = (String) this.elementResolver.resolve(
                tagTokens.get(0), bindingDataList);
        String[] parts = resolvedNetwork.split("/");
        Inet4Address networkAddress;
        try {
            networkAddress = (Inet4Address) InetAddress.getByName(parts[0]);
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unable to get the network address " + parts[0], e);
        }
        ByteBuffer fourBytes = ByteBuffer.wrap(new byte[4]);
        fourBytes.put(networkAddress.getAddress());
        fourBytes.rewind();
        int network = fourBytes.getInt();
        int index = Integer.parseInt(this.elementResolver.resolve(
                tagTokens.get(1), bindingDataList).toString());
        int prefix = Integer.parseInt(parts[1]);
        if (index == 0 || index >= (1 << (32 - prefix))) {
            throw new TemplateEngineException(index + " is out of ip range 2^" + prefix);
        }
        int ipAddressInt = network + index;
        fourBytes.rewind();
        fourBytes.putInt(ipAddressInt);
        Inet4Address ipAddress;
        try {
            ipAddress = (Inet4Address) InetAddress.getByAddress(fourBytes.array());
        } catch (UnknownHostException e) {
            throw new TemplateEngineException("Unabled to get ip address " + fourBytes, e);
        }
        return ipAddress.getHostAddress();
    }
}

File: templating/tag_contributions/src/main/go/src/jsonteng_contribs/tags/ipv4_host_ip_tag.go
// Copyright 2019 VMware, Inc.
// SPDX-License-Indentifier: Apache-2.0

package tags

import (
	"container/list"
	"fmt"
	"jsonteng/errors"
	intfs "jsonteng/interfaces"
	"net"
	"strconv"
)

type ipv4HostIPTag struct {
	intfs.TagBase
	elementResolver intfs.ElementResolver
}

func (tag *ipv4HostIPTag) init(tagResolver intfs.TagResolver) {
	tag.Name = "ipv4-host-ip"
	tag.TagResolver = tagResolver
	tag.elementResolver = tagResolver.GetElementResolver()
}

func (tag *ipv4HostIPTag) Process(tagTokens []interface{}, bindingDataList *list.List) (interface{}, error) {
	tokenCount := len(tagTokens)
	if tokenCount != 2 {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Tag \"%s\" requires 2 parameters. Parameters given %v", tag.Name, tagTokens)
		return nil, &errl
	}
	resolvedNetworkObj, err := tag.elementResolver.ResolveElement(tagTokens[0], bindingDataList)
	if err != nil {
		return nil, err
	}
	resolvedNetwork, ok := resolvedNetworkObj.(string)
	if !ok {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Invalid network %v", resolvedNetworkObj)
		return nil, &errl
	}
	_, ipv4Net, err := net.ParseCIDR(resolvedNetwork)
	if err != nil {
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Invalid network format %v", resolvedNetwork)
		return nil, &errl
	}
	network := ipv4Net.IP
	indexObj, err := tag.elementResolver.ResolveElement(tagTokens[1], bindingDataList)
	if err != nil {
		return nil, err
	}
	var index int64
	switch indexTemp := indexObj.(type) {
	case float64:
		index = int64(indexTemp)
	case string:
		index, err = strconv.ParseInt(indexTemp, 10, 32)
		if err != nil {
			errl := errors.GenericError
			errl.Message = fmt.Sprintf("Invalid address index %v", indexTemp)
			return nil, &errl
		}
	default:
		errl := errors.GenericError
		errl.Message = fmt.Sprintf("Unknown address index type %v", indexTemp)
		return nil, &errl
	}
	return net.IPv4(network[0], network[1], network[2], network[3]+byte(index)).String(), nil
}
